lib/plugins/aws/package/compile/events/cognitoUserPool/index.js

```diff
@@ -108,56 +108,78 @@ class AwsCompileCognitoUserPoolEvents {
-            const customCognitoUserPoolResourceLogicalId = this.provider.naming.getCustomResourceCognitoUserPoolResourceLogicalId(
-              functionName,
-              idx
+            const customPoolResourceLogicalId = this.provider.naming.getCustomResourceCognitoUserPoolResourceLogicalId(
+              functionName
             );
 
-            const customCognitoUserPool = {
-              [customCognitoUserPoolResourceLogicalId]: {
-                Type: 'Custom::CognitoUserPool',
-                Version: 1.0,
-                DependsOn: [eventFunctionLogicalId, customResourceFunctionLogicalId],
-                Properties: {
-                  ServiceToken: {
-                    'Fn::GetAtt': [customResourceFunctionLogicalId, 'Arn'],
-                  },
-                  FunctionName,
-                  UserPoolName: pool,
-                  UserPoolConfig: {
-                    Trigger: trigger,
+            // store how often the custom Cognito User Pool resource is used
+            if (poolResources[pool]) {
+              poolResources[pool] = _.union(poolResources[pool], [customPoolResourceLogicalId]);
+            } else {
+              Object.assign(poolResources, {
+                [pool]: [customPoolResourceLogicalId],
+              });
+            }
+
+            let customCognitoUserPoolResource;
+            if (numEventsForFunc === 1) {
+              customCognitoUserPoolResource = {
+                [customPoolResourceLogicalId]: {
+                  Type: 'Custom::CognitoUserPool',
+                  Version: 1.0,
+                  DependsOn: [eventFunctionLogicalId, customResourceFunctionLogicalId],
+                  Properties: {
+                    ServiceToken: {
+                      'Fn::GetAtt': [customResourceFunctionLogicalId, 'Arn'],
+                    },
+                    FunctionName,
+                    UserPoolName: pool,
+                    UserPoolConfigs: [
+                      {
+                        Trigger: trigger,
+                      },
+                    ],
                   },
                 },
-              },
-            };
-
-            _.merge(compiledCloudFormationTemplate.Resources, customCognitoUserPool);
-
-            iamRoleStatements.push({
-              Effect: 'Allow',
-              Resource: '*',
-              Action: [
-                'cognito-idp:ListUserPools',
-                'cognito-idp:DescribeUserPool',
-                'cognito-idp:UpdateUserPool',
-              ],
-            });
+              };
+
+              iamRoleStatements.push({
+                Effect: 'Allow',
+                Resource: '*',
+                Action: [
+                  'cognito-idp:ListUserPools',
+                  'cognito-idp:DescribeUserPool',
+                  'cognito-idp:UpdateUserPool',
+                ],
+              });
+            } else {
+              Resources[customPoolResourceLogicalId].Properties.UserPoolConfigs.push({
```

Comment:
If I read correctly, logic as put doesn't handle well configuration where user configured two pools under one function, and will assign two triggers to first approached pool (?) e.g. following:

```yaml
functions:
  users:
    handler: users.handler
    events:
      - cognitoUserPool:
          pool: legacy-user-pool
          trigger: CustomMessage
          existing: true
      - cognitoUserPool:
          pool: other-pool
          trigger: OtherMessage
          existing: true
```

will confgure `CustomMessage` and `OtherMessage` triggers for `legacy-user-pool`.

I think we should rather throw here if we approach events configuration for two different pools.

To ease the validation and make configuration schema less error-prone, maybe it'll be better to not allow more than one `cognitoUserPool` event per function, and allow to list triggers with a list, e.g.:

```yaml
functions:
  users:
    handler: users.handler
    events:
      - cognitoUserPool:
          pool: legacy-user-pool
          trigger:
            - CustomMessage
            - PreSignUp
          existing: true
```

Reply:
Good point. I already added a note in the docs and will add a check for it here as well.

Unfortunately we cannot alter the event definition since this has to be aligned with the event definition we put in place when implementing Cognito User Pool event support for non-existing pools.